### PR TITLE
Roles, token authentication, and cypher query service

### PR DESCRIPTION
**Pull request by jar398**
There's still a lot I don't get about rails. I've put what I think is way too much logic into a controller class (services_controller.rb) but I don't know how to structure it better. This and many other defects are probably not fatal and can be fixed later (technical debt).

### CODE REVIEW

File: app/controllers/service/cypher_controller.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+      cypher = params[:query]
+    else
+      # default for testing.  remove this before release.
+      cypher = "MATCH (n:Page) RETURN n.page_id LIMIT 10;"
+    end
+
+    # Web users are not supposed to modify the database.  The following
+    # is an ad hoc filter to try to prevent this, at least when it is 
+    # accidental; one shouldn't expect it to be secure against a concerted attack.
+    if cypher =~ /\w(delete|create|set|remove|merge|call|drop|load)\w/i
```
Comment (JRice) [inline app/controllers/service/cypher_controller.rb]: Errr... I *think* you intended to use `\b` here instead of `\w`. E.g.:

```
DEVELOPMENT [8] pry(main)> cypher = "this has the word create in it"
=> "this has the word create in it"
DEVELOPMENT [9] pry(main)> cypher =~ /\w(delete|create|set|remove|merge|call|drop|load)\w/i
=> nil
DEVELOPMENT [10] pry(main)> cypher =~ /\b(delete|create|set|remove|merge|call|drop|load)\b/i
=> 18
```

File: app/controllers/service/cypher_controller.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+      render status: 403    #403 = forbidden (401 means unauthenticated)
+    else
+      render json: TraitBank.query(cypher)
```
Comment (JRice) [inline app/controllers/service/cypher_controller.rb]: I wonder, and this is purely speculative, but I do wonder... whether we should return the results of the call to #query, which is formatted more for neography (a gem) and includes some ancilary information, or if we should automatically (assuming the query was successful) return `TraitBank.query(cypher)['data']` (with some extra fluff in there to handle nil results, of course). Currently it will return something like this:

```
{"columns"=>["n.page_id"], "data"=>[[351759], [351760]]}
```

...And *perhaps* (not sure) all they want is the value of "data", here. I'm probably wrong. It might be worth expressing the nature of the return value in the docs somewhere, is all. [shrug] Ignore this comment, it's not actionable :D ...but I'm saving it for future discussion.

File: app/controllers/services_controller.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  # No args - act as a guard on any web service action that wants authentication
+
+  def authorize_user_from_token!
+    if current_user
+      raise "Forbidden" unless @current_user.is_power_user?    # Unauthorized
+      @current_user = current_user
+    else
+      the_claims = claims
+      raise "Unauthenticated" unless the_claims    # Unauthenticated
+      if user = User.find_by(email: the_claims[0]['user']) and
```
Comment (JRice) [inline app/controllers/services_controller.rb]: Style: Prefer `&&` unless there is a strong reason to use `and`. [[source](https://github.com/rubocop-hq/ruby-style-guide#user-content-no-and-or-or)]

Hmmmn. Also, I would alter `User.find_by(email:  foo)`, which returns an array (also, by the way but no important now, `#where` is preferred, stylistically), to `User.find_by_email(foo)`, which returns a single instance (the first), and thus you could avoid using `[0]` on line 35 below.

File: app/controllers/services_controller.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        raise "Forbidden" unless user.is_power_user?    # Unauthorized
+        @current_user = user
+      end
+    end
+  end
+
+  # Aux
+  def claims
+    auth_header = request.headers['Authorization'] and
+      token = auth_header.split(' ').last and
```
Comment (JRice) [inline app/controllers/services_controller.rb]: AFAIK, split's arugment defaults to `' '`, so I think you could just say `auth_header.split.last` here.

File: app/controllers/services_controller.rb [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+      ::TokenAuthentication.decode(token)
+      # I don't know what the double-colon means
```
Comment (JRice) [inline app/controllers/services_controller.rb]: It means "don't look for this class as a child of the *current* class, but as a "top-level" class. You *probably* added that because without it, it was looking for `ServicesController::TokenAuthentication`. ...But you might not need the `::` if you've created a file called `token_authentication.rb` in the load path. ...I will say, I've seen Ruby (Rails, really) be quite picky about this, so I don't mind seeing it here. It's not an uncommon syntax.

File: app/controllers/services_controller.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  rescue
```
Comment (JRice) [inline app/controllers/services_controller.rb]: A "naked" rescue is quite broad, and will catch ALL errors, so if there were, say, a syntax error added to the code, we might never discover it. It's best to "know" which errors you expect and rescue them specifically, if you know they are benign. This code would be considered "dangerous" as is (in terms of breaking functionality down the line).

File: app/models/token_authentication.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+require 'jwt'
+
+# Not an ActiveRecord; not instantiated, actually
```
Comment (JRice) [inline app/models/token_authentication.rb]: The standard practice, if you will never instantiate a class, is to make it a `module` and then `extend` (for class methods) or `include` (for instance methods) your module as needed. So in this case, you could call `TokenAuthentication` a `module` and, above in `ServicesController`, you could `extend TokenAuthentication`, and you should have access to these two class methods "natively" (no need to call TokenAuthentication specifically).

Also, when you have more than one class method being declared, it is stylistically preferable to use this syntax:
```ruby
class Foo
  class << self
    def first_class_method
      # whatever
    end
    def second_class_method
      # more stuff
    end
  end
end
```

...instead of the `def self.method` syntax.

File: db/migrate/20180701234758_add_role_to_users.rb [hunk 8]
```diff
@@ -0,0 +1,24 @@
+class AddRoleToUsers < ActiveRecord::Migration
+
+  # 0 = nobody, 100 = admin.  See enum in models/user.rb
+  # There must be a better way to do this
+
+  @@user = 10
```
Comment (JRice) [inline db/migrate/20180701234758_add_role_to_users.rb]: You can actually get this right off of the class. i.e.: `User.roles[:user]`. ...You could also avoid your custom integers that way!

File: db/migrate/20180701234758_add_role_to_users.rb [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  @@admin = 100
+
+  def up
+    add_column :users, :role, :integer, null: false, default: @@user
+    User.connection.execute("UPDATE users SET role = #{@@admin} WHERE admin = true")
```
Comment (JRice) [inline db/migrate/20180701234758_add_role_to_users.rb]: THANK YOU. :) I was watching for this.  ;)  Of course, you can do it more elegantly with `User.where(admin: true).update_all(role: User.roles[:admin])` ... but still!

File: app/controllers/service/cypher_controller.rb (hunk 1 above)
Comment (jar398) [reply app/controllers/service/cypher_controller.rb]: yep

File: app/controllers/service/cypher_controller.rb (hunk 2 above)
Comment (jar398) [reply app/controllers/service/cypher_controller.rb]: My idea was to make the cypher service universal, implementing raw cypher, whose format someone can read about on the neo4j site. Also some scripts processing the output will need to know the column names, and it's easier to get them from "columns" than from the query itself. For general use we'll have a whole suite of specific queries for particular purposes, and the output will be reformatted to look more like json-ld (maybe even to _be_ json-ld), and I expect to phase out general use of the /cypher service.

File: app/controllers/services_controller.rb (hunk 3 above)
Comment (jar398) [reply app/controllers/services_controller.rb]: re `&&`, ok, I don't understand this, but will accept without question. I was just copying other code I think.
re `find_by_email`, didn't know about that. will switch

File: app/controllers/services_controller.rb (hunk 4 above)
Comment (jar398) [reply app/controllers/services_controller.rb]: copied this practice from the interwebs. a ruby novice like me would find it clearer, but will change

File: app/controllers/services_controller.rb (hunk 6 above)
Comment (jar398) [reply app/controllers/services_controller.rb]: hmm.  I guess I will have to read up on `rescue`.  again, i copied this construct from the interwebs (sorry!! I know that's not right but I get fatigued reading so much documentation)

File: app/models/token_authentication.rb (hunk 7 above)
Comment (jar398) [reply app/models/token_authentication.rb]: So what directory is this utility supposed to live in? One source suggested creating a `lib/` directory but I didn't want to disrupt the current structure.

And `ServicesController` is already a subclass of `ApplicationController` - how can it extend two things?...  I guess I should read up on ruby classes and modules although so far I've only seen tutorial like presentations, nothing that actually explains how they work.

I've seen that class-in-class (singleton?) business before but don't want to use it until I understand it.

File: db/migrate/20180701234758_add_role_to_users.rb (hunk 8 above)
Comment (jar398) [reply db/migrate/20180701234758_add_role_to_users.rb]: OK. (The scoping & visibility rules remain a mystery for me.)

File: app/controllers/service/cypher_controller.rb (hunk 2 above)
Comment (jar398) [reply app/controllers/service/cypher_controller.rb]: By the way if there's an error in parsing or otherwise processing the query, the diagnostic output should be reflected back in the 500 (or 400) response (currently just shows up on the server console). That's on my to-do list

File: app/models/token_authentication.rb (hunk 7 above)
Comment (JRice) [reply app/models/token_authentication.rb]: There are two schools of thought for where to put files. The older (and arguably more popular one) is that `/app/models` is for relational models stored in the database, and `/lib` is for everything else. The other, which I prefer and am using here, is that `/app/models` are for classes specific to your Rails project, and `/lib` is for more general classes that could (really "should") be extracted into their own projects. [shrug] It's not a huge problem to use one or the other.

Right, `ServicesController` is already a subclass of `ApplicationController`. It will contnue to be. the use of `extend` does NOT imply inheritance (e.g.: subclassing) of any kind. `extend` means, roughly, "steal the class methods (and variables) from this other object". If you're a Java sort, you can think of a module as a kind of abstract class and `extend` as some version of `implements`. ...Kind of. :) Ruby is a little more loosey-goosey in this domain. :D

There are no singletons involved, here, it's just a way of referencing methods from another class. ...A separation of concerns.

TL;DR: Calling it a "class" instead of a "module" isn't strictly *wrong*, but stylistically it's preferable to make it a `module` and to use `extend`. Don't sweat it if the change is intimidating.

File: db/migrate/20180701234758_add_role_to_users.rb (hunk 9 above)
Comment (jar398) [reply db/migrate/20180701234758_add_role_to_users.rb]: I tried rewriting the sql commands as you suggested but couldn't come up with something that I had confidence in. So I've left these alone - they will only be run once per installation anyhow.

File: app/controllers/service/cypher_controller.rb (hunk 2 above)
Comment (jar398) [reply app/controllers/service/cypher_controller.rb]: Oh I see, you're talking about the case where the query defaults? I should probably remove it, it's only there for debugging (so you don't have to always come up with a query when testing)
